File: Editorial/126_Word_Ladder_II/Python3/2_Bidirectional_Breadth_First_Search_Backtracking.py
class Solution:
    def __init__(self):
        self.adjList = {}
        self.currPath = []
        self.shortestPaths = []

    def findNeighbors(self, word, wordList):
        neighbors = []
        charList = list(word)
        for i in range(len(word)):
            oldChar = charList[i]
            for ch in range(97, 123):
                c = chr(ch)
                charList[i] = c
                if c == oldChar or "".join(charList) not in wordList:
                    continue
                neighbors.append("".join(charList))
            charList[i] = oldChar
        return neighbors

    def backtrack(self, source, destination):
        if source == destination:
            tempPath = list(self.currPath)
            self.shortestPaths.append(tempPath)
        for i in range(len(self.adjList.get(source, []))):
            self.currPath.append(self.adjList[source][i])
            self.backtrack(self.adjList[source][i], destination)
            self.currPath.pop()

    def addEdge(self, word1, word2, direction):
        if direction == 1:
            self.adjList[word1] = self.adjList.get(word1, []) + [word2]
        else:
            self.adjList[word2] = self.adjList.get(word2, []) + [word1]

    def bfs(self, beginWord, endWord, wordList):
        if endWord not in wordList:
            return False
        if beginWord in wordList:
            wordList.remove(beginWord)
        forwardQueue = set([beginWord])
        backwardQueue = set([endWord])
        found = False
        direction = 1
        while len(forwardQueue) != 0:
            visited = set()
            if len(forwardQueue) > len(backwardQueue):
                forwardQueue, backwardQueue = backwardQueue, forwardQueue
                direction ^= 1
            for currWord in forwardQueue:
                neighbors = self.findNeighbors(currWord, wordList)
                for word in neighbors:
                    if word in backwardQueue:
                        found = True
                        self.addEdge(currWord, word, direction)
                    elif (
                        not found
                        and word in wordList
                        and word not in forwardQueue
                    ):
                        visited.add(word)
                        self.addEdge(currWord, word, direction)
            for currWord in forwardQueue:
                if currWord in wordList:
                    wordList.remove(currWord)
            if found:
                break
            forwardQueue = visited
        return found

    def findLadders(self, beginWord, endWord, wordList):
        copiedWordList = set(wordList)
        sequence_found = self.bfs(beginWord, endWord, copiedWordList)
        if sequence_found == False:
            return self.shortestPaths
        self.currPath.append(beginWord)
        self.backtrack(beginWord, endWord)
        return self.shortestPaths

File: Editorial/0101-0200/0126_Word_Ladder_II/CSharp/2_Bidirectional_Breadth_First_Search_Backtracking.cs
public class Solution {
    Dictionary<string, List<string>> adjList =
        new Dictionary<string, List<string>>();

    List<string> currPath = new List<string>();
    List<List<string>> shortestPaths = new List<List<string>>();

    private List<string> findNeighbors(string word, HashSet<string> wordList) {
        List<string> neighbors = new List<string>();
        char[] charList = word.ToCharArray();
        for (int i = 0; i < word.Length; i++) {
            char oldChar = charList[i];
            for (char c = 'a'; c <= 'z'; c++) {
                charList[i] = c;
                if (c == oldChar || !wordList.Contains(new String(charList))) {
                    continue;
                }

                neighbors.Add(new String(charList));
            }

            charList[i] = oldChar;
        }

        return neighbors;
    }

    private void backtrack(string source, string destination) {
        if (source.Equals(destination)) {
            List<string> tempPath = new List<string>(currPath);
            shortestPaths.Add(tempPath);
        }

        if (!adjList.ContainsKey(source)) {
            return;
        }

        for (int i = 0; i < adjList[source].Count; i++) {
            currPath.Add(adjList[source][i]);
            backtrack(adjList[source][i], destination);
            currPath.RemoveAt(currPath.Count - 1);
        }
    }

    private void addEdge(string word1, string word2, int direction) {
        if (direction == 1) {
            if (!adjList.ContainsKey(word1)) {
                adjList[word1] = new List<string>();
            }

            adjList[word1].Add(word2);
        } else {
            if (!adjList.ContainsKey(word2)) {
                adjList[word2] = new List<string>();
            }

            adjList[word2].Add(word1);
        }
    }

    private bool bfs(string beginWord, string endWord,
                     HashSet<string> wordList) {
        if (!wordList.Contains(endWord)) {
            return false;
        }

        if (wordList.Contains(beginWord)) {
            wordList.Remove(beginWord);
        }

        HashSet<string> forwardQueue = new HashSet<string>();
        HashSet<string> backwardQueue = new HashSet<string>();
        forwardQueue.Add(beginWord);
        backwardQueue.Add(endWord);
        bool found = false;
        int direction = 1;
        while (forwardQueue.Count != 0) {
            HashSet<string> visited = new HashSet<string>();
            if (forwardQueue.Count > backwardQueue.Count) {
                HashSet<string> temp = forwardQueue;
                forwardQueue = backwardQueue;
                backwardQueue = temp;
                direction ^= 1;
            }

            foreach (string currWord in forwardQueue) {
                List<string> neighbors = findNeighbors(currWord, wordList);
                foreach (string word in neighbors) {
                    if (backwardQueue.Contains(word)) {
                        found = true;
                        addEdge(currWord, word, direction);
                    } else if (!found && wordList.Contains(word) &&
                               !forwardQueue.Contains(word)) {
                        visited.Add(word);
                        addEdge(currWord, word, direction);
                    }
                }
            }

            foreach (string currWord in forwardQueue) {
                if (wordList.Contains(currWord)) {
                    wordList.Remove(currWord);
                }
            }

            if (found) {
                break;
            }

            forwardQueue = visited;
        }

        return found;
    }

    public IList<IList<string>> FindLadders(string beginWord, string endWord,
                                            IList<string> wordList) {
        HashSet<string> copiedWordList = new HashSet<string>(wordList);
        bool sequence_found = bfs(beginWord, endWord, copiedWordList);
        if (sequence_found == false) {
            return shortestPaths.ToArray();
        }

        currPath.Add(beginWord);
        backtrack(beginWord, endWord);
        return shortestPaths.ToArray();
    }
}
